File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/audit.ts
/**
 * Deterministic audit log and export summary for campaign state. Both
 * endpoints report counts derived directly from SQLite; no data is mutated.
 */

import { getDb } from '../db.ts';
import { SCHEMA_VERSION } from '../db.ts';
import type { ApiResult } from '../types.ts';

interface CampaignRow {
  id: string;
  name: string;
}

function countRows(table: string, campaignId: string): number {
  const db = getDb();
  const row = db.prepare(`SELECT COUNT(*) as count FROM ${table} WHERE campaign_id = ?`).get(campaignId) as {
    count: number;
  };
  return row.count;
}

export function campaignAudit(campaignId: string): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId) as CampaignRow | undefined;
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  return {
    status: 200,
    body: {
      campaign_id: campaignId,
      events: countRows('campaign_events', campaignId),
      quests: countRows('campaign_quests', campaignId),
      npcs: countRows('campaign_npcs', campaignId),
      sessions: countRows('campaign_sessions', campaignId),
    },
  };
}

export function campaignExport(campaignId: string): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id, name FROM campaigns WHERE id = ?').get(campaignId) as
    | CampaignRow
    | undefined;
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  return {
    status: 200,
    body: {
      campaign_id: campaign.id,
      name: campaign.name,
      characters: countRows('campaign_characters', campaignId),
      quests: countRows('campaign_quests', campaignId),
      npcs: countRows('campaign_npcs', campaignId),
      inventory_items: countRows('campaign_inventory', campaignId),
      sessions: countRows('campaign_sessions', campaignId),
      schema_version: SCHEMA_VERSION,
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/audit.rb
# Deterministic audit log and export summary for campaign state.

get '/v1/campaigns/:campaign_id/audit' do
  halt 404, { error: 'unknown campaign' }.to_json unless campaign_exists?(params[:campaign_id])

  events = db.execute('SELECT COUNT(*) AS cnt FROM campaign_events WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  quests = db.execute('SELECT COUNT(*) AS cnt FROM campaign_quests WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  npcs = db.execute('SELECT COUNT(*) AS cnt FROM campaign_npcs WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  sessions = db.execute('SELECT COUNT(*) AS cnt FROM campaign_sessions WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']

  {
    campaign_id: params[:campaign_id],
    events: events,
    quests: quests,
    npcs: npcs,
    sessions: sessions
  }.to_json
end

get '/v1/campaigns/:campaign_id/export' do
  campaign = db.execute('SELECT * FROM campaigns WHERE id = ?', [params[:campaign_id]]).first
  halt 404, { error: 'unknown campaign' }.to_json unless campaign

  characters = db.execute('SELECT COUNT(*) AS cnt FROM campaign_characters WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  quests = db.execute('SELECT COUNT(*) AS cnt FROM campaign_quests WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  npcs = db.execute('SELECT COUNT(*) AS cnt FROM campaign_npcs WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  inventory_items = db.execute('SELECT COUNT(DISTINCT item_slug) AS cnt FROM campaign_inventory WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  sessions = db.execute('SELECT COUNT(*) AS cnt FROM campaign_sessions WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']

  {
    campaign_id: campaign['id'],
    name: campaign['name'],
    characters: characters,
    quests: quests,
    npcs: npcs,
    inventory_items: inventory_items,
    sessions: sessions,
    schema_version: SCHEMA_VERSION
  }.to_json
end
